cmds/switch_root/switch_root.go

```diff
@@ -23,140 +24,201 @@ func usage() string {
+func isSymlink(mode uint32) bool {
+	// numbers come from `man fstatat`
+	return (mode & 0170000) == 0120000
+}
+
+// recursively deletes a directory and everything in it
+// works with a file descriptor, can delete files not referenceable
+// any more (e.g. after chroot)
+// does not descent into mounts
+func recursiveDelete(fd int) error {
+	var rb syscall.Stat_t
+
+	if err := syscall.Fstat(fd, &rb); err != nil {
+		return err
 	}
 
-	if fileStat, err := file.Stat(); fileStat.IsDir() {
-		if err != nil {
-			return err
-		}
+	// may need to call syscall.ReadDirent multiple times
+	for {
 
-		names, err := file.Readdirnames(-1)
-		defer file.Close()
+		// get the subfiles in buff
+		buff := make([]byte, 4096)
+		var nbuff int
+		nbuff, err := syscall.ReadDirent(fd, buff)
 		if err != nil {
 			return err
 		}
-
-		for _, fileName := range names {
-
-			if fileName == "." || fileName == ".." {
-				continue
-			}
-
-			if err := littleDoctor(filepath.Join(path, fileName), fsType); err != nil {
-				return err
-			}
-			if err := os.Remove(path); err != nil {
-				return err
-			}
+		if nbuff <= 0 {
+			break
 		}
 
-	} else {
-		if err := os.Remove(path); err != nil {
-			return err
+		_, _, names := syscall.ParseDirent(buff, nbuff, []string{})
+
+		for _, name := range names {
+			// try to open it with O_DIRECTORY to know if it's a directory
+			// NOTE O_DIRECTORY is linux specific
+			namefd, err := syscall.Openat(fd, name, syscall.O_DIRECTORY, syscall.O_RDWR)
+			if err != nil {
+				// just delete files
+				if err := unlinkat(fd, name, 0); err != nil {
+					return err
+				}
+			} else {
+				defer syscall.Close(namefd)
+
+				var nameStatT syscall.Stat_t
+				if err := fstatat(fd, name, &nameStatT, unix.AT_SYMLINK_NOFOLLOW); err != nil {
+					return err
+				}
+
+				if fdDev, err := getDev(fd); err != nil {
+					return err
+				} else if fdDev != nameStatT.Dev {
+					// on different filesystem, don't go there
+					continue
+				}
+
+				// not actually a dir, but a symlink to a dir, treat like file and remove
+				if isSymlink(nameStatT.Mode) {
+					if err := unlinkat(fd, name, 0); err != nil {
+						return err
+					}
+					continue
+				}
+
+				// actual recursion
+				if err := recursiveDelete(namefd); err != nil {
+					return err
+				}
+				// back from recursion, dir is not empty, delete
+				if err := unlinkat(fd, name, AT_REMOVEDIR); err != nil {
+					return err
+				}
+			}
 		}
 	}
-
 	return nil
 }
 
-// execCommand will run the executable at "path" with PID 1.
+// do a proper exec to retain PID 1.
 // It returns an error if the command exits incorrectly.
 func execCommand(path string) error {
-	cmd := exec.Command(path)
-	cmd.Stdin = os.Stdin
-	cmd.Stderr = os.Stderr
-	cmd.Stdout = os.Stdout
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Setctty:    true,
-		Setsid:     true,
-		Cloneflags: syscall.CLONE_THREAD | syscall.CLONE_NEWPID,
-	}
-
-	return cmd.Run()
+	return syscall.Exec(path, []string{path}, []string{})
```

Comment:
Define "not work well"? I think CLONE_THREAD was wrong in the old code -- if you delete that, execCommand should work as it should.

This doesn't seem to pass CLONE_NEWPID either, btw?

Reply:
Oh nevermind... there is no new child. Just an exec. That works.

Reply:
Yeah, honestly I am not 100% sure of the implications of using `CLONE_NEWPID` here, but I want to avoid trouble down the line when launching systemd because it is somehow not the 'real' PID 1. Instead, we decided to stay close to the reference impl in `util-linux` here and do a real exec. (I obviously intend to run this as PID 1, so it keeps that for systemd)

Reply:
I use CLONE_NEWPID in  cmds/init/init.go and a regular os.Exec and it works fine -- the child gets its own PID name space and thinks it's pid 1 and all is well. Just a data point.

Reply:
I would recommend not using CLONE_NEWPID here. While upstream doesn't specify this clearly 
https://www.freedesktop.org/wiki/Software/systemd/InitrdInterface/ 
https://www.kernel.org/doc/Documentation/filesystems/ramfs-rootfs-initramfs.txt
(it just mentions exec/execute without referring to an actual function), pretty much every implementation of switch_root uses `execv()` to jump into pid 1:
https://github.com/karelzak/util-linux/blob/master/sys-utils/switch_root.c#L260
https://git.busybox.net/busybox/tree/util-linux/switch_root.c#n286
https://github.com/landley/toybox/blob/master/toys/other/switch_root.c#L112
CLONE_NEWPID might seem to work on a first look, but it's not something that either upstream (kernel and systemd) will test against, and it could lead to subtle hard to debug issues later down the road. I could also this potentially confusing systemd container detection logic (as containers are often implemented with CLONE_NEWPID).

Reply:
Any docker container that runs with a full init system would have the same problem, but I agree that execv will work just fine here and there's no need to make it more complicated than it needs to be.